### PR TITLE
[SPARK-26543][SQL] Support the coordinator to determine post-shuffle partitions more reasonably

### PR DESCRIPTION
**Pull request by southernriver**
## What changes were proposed in this pull request?

For SparkSQL ,when we open AE by 'set spark.sql.adapative.enable=true'，the ExchangeCoordinator will introduced to determine the number of post-shuffle partitions. But in some certain conditions,the coordinator performed not very well, there are always some tasks retained and they worked with Shuffle Read Size / Records 0.0B/0 ,We could increase the spark.sql.adaptive.shuffle.targetPostShuffleInputSize to solve this,but this action is unreasonable as targetPostShuffleInputSize Should not be set too large. As follow:
image

We could   reproduce this problem easily with the SQL:

`set spark.sql.adaptive.enabled=true；`
`  spark.sql.shuffle.partitions 100；`
`  spark.sql.adaptive.shuffle.targetPostShuffleInputSize  33554432 ；`
`  SELECT a,COUNT(1) FROM TABLE  GROUP BY  a DISTRIBUTE BY cast(rand()* 10 as bigint)` 

before fix：
image
after fix：
image

## How was this patch tested?
manual and  unit tests

### CODE REVIEW

**Comment by southernriver** [comment]
@cloud-fan @carsonwang  Could you please have a look at this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ExchangeCoordinator.scala [hunk 1]
```diff
@@ -178,11 +178,17 @@ class ExchangeCoordinator(
...
+          partitionStartIndices += i
+        }
         // reset postShuffleInputSize.
         postShuffleInputSize = nextShuffleInputSize
       } else postShuffleInputSize += nextShuffleInputSize
 
+      // filter the last indice which will split the postShuffleInputSize=0
+      if (postShuffleInputSize == 0 && i == numPreShufflePartitions - 1) {
+        partitionStartIndices.remove(partitionStartIndices.size - 1)
+      }
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ExchangeCoordinator.scala]: Can you move this check outside the `while` loop?

**Comment by southernriver** [comment]
@maropu  fix that ,thank you.

**Comment by southernriver** [comment]
@dongjoon-hyun @maropu  This patch merges cleanly before this, but today it indicates that   "This branch has conflicts that must be resolved", Could you please have a look at this?

**Comment by maropu** [comment]
You need to resolve the conflict by a git-rebase command.

**Comment by cloud-fan** [comment]
@southernriver can you check if this problem still exists in the new AQE framework?

**Pull request closed without merging**